File: modules/01e-trees-ensembles/ts/02-random-forest.ts
/**
 * Task 2 🟡 — Bagging → random forest.
 *
 * What you'll learn:
 *   - Bootstrap sampling: draw n indices WITH replacement — each resample
 *     leaves out ≈ 36.8% of the rows (the fraction of unique rows → 1 - 1/e
 *     ≈ 63.2%).
 *   - Bagging: train one tree per bootstrap, average (majority-vote) their
 *     predictions — averaging B noisy estimators divides the uncorrelated
 *     part of their variance by B.
 *   - The random-forest trick: at EVERY split, consider only a random subset
 *     of maxFeatures features, so the trees stop all making the same greedy
 *     first split and become decorrelated — which is what makes the
 *     averaging work.
 *
 * The math (README derives each step):
 *
 *   Unique-fraction of a bootstrap:  P(row i never drawn in n tries)
 *       = (1 - 1/n)^n → e⁻¹ ≈ 0.368,  so  unique ≈ 63.2% of rows.
 *
 *   Variance of an average of B estimators with variance σ² and pairwise
 *   correlation ρ:
 *       Var( (1/B) Σ f_b ) = ρσ² + (1-ρ)σ²/B
 *   Bagging shrinks the second term; feature subsampling shrinks ρ — the
 *   first.
 *
 * You implement: bootstrapSample, trainForest, forestPredict. A full
 * single-tree trainer that accepts maxFeatures (the per-split feature
 * sampler), the dataset (2 signal + 3 pure-noise features), and the harness
 * comparing individual trees vs the ensemble vs a single deep-tree baseline
 * are provided and runnable.
 *
 * How to run:
 *   pnpm tsx modules/01e-trees-ensembles/ts/02-random-forest.ts
 */

const SEED = 21;
const N = 520; // total samples (first N_TRAIN train, rest test)
const N_TRAIN = 370;
const NOISE_FLIP = 0.1;
const N_TREES = 25;
const MAX_FEATURES = 2; // features considered per split (out of 5)

// ---------------------------------------------------------------------------
// Seeded RNG (provided) — LCG + Box-Muller for reproducible randomness
// ---------------------------------------------------------------------------

function makeRng(seed: number): () => number {
  let s = seed >>> 0;
  return () => {
    // Numerical Recipes LCG
    s = (Math.imul(s, 1664525) + 1013904223) >>> 0;
    return s / 4294967296; // uniform in [0, 1)
  };
}

function makeGaussian(seed: number): () => number {
  const u = makeRng(seed);
  return () => {
    // Box-Muller: two uniforms → one standard normal
    const u1 = u() + 1e-12;
    const u2 = u();
    return Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
  };
}

// ---------------------------------------------------------------------------
// Tree node type + synthetic data (provided — do not edit)
// ---------------------------------------------------------------------------

type TreeNode =
  | { leaf: true; prediction: number }
  | {
      leaf: false;
      feature: number;
      threshold: number;
      left: TreeNode;
      right: TreeNode;
    };

/**
 * 5 features: x0, x1 carry the signal (the offset-XOR pattern from Task 1);
 * x2..x4 are pure Gaussian noise — bait for an overfitting deep tree.
 * NOISE_FLIP of the labels are flipped.
 */
function makeData(): {
  XTrain: number[][];
  yTrain: number[];
  XTest: number[][];
  yTest: number[];
} {
  const u = makeRng(SEED);
  const g = makeGaussian(SEED + 50);
  const X = Array.from({ length: N }, () => [
    -3 + 6 * u(),
    -3 + 6 * u(),
    g(),
    g(),
    g(),
  ]);
  const flip = makeRng(SEED + 99);
  const y = X.map((row) => {
    const label = row[0] > 0.8 !== row[1] > -0.7 ? 1 : 0;
    return flip() < NOISE_FLIP ? 1 - label : label;
  });
  return {
    XTrain: X.slice(0, N_TRAIN),
    yTrain: y.slice(0, N_TRAIN),
    XTest: X.slice(N_TRAIN),
    yTest: y.slice(N_TRAIN),
  };
}

// ---------------------------------------------------------------------------
// Single-tree trainer (provided — do not edit)
//
// The same CART you built in Task 1, with ONE addition: if maxFeatures is
// set, each split only scans a random subset of that many features (drawn
// with the rng) — the random-forest decorrelation trick.
// ---------------------------------------------------------------------------

function giniImpurity(y: number[]): number {
  if (y.length === 0) return 0;
  let ones = 0;
  for (const v of y) ones += v;
  const p1 = ones / y.length;
  return 1 - ((1 - p1) ** 2 + p1 ** 2);
}

function majorityClass(y: number[]): number {
  let ones = 0;
  for (const v of y) ones += v;
  return ones * 2 > y.length ? 1 : 0;
}

function pickFeatures(
  rng: () => number,
  d: number,
  maxFeatures: number | null,
): number[] {
  const all = Array.from({ length: d }, (_, j) => j);
  if (maxFeatures === null || maxFeatures >= d) return all;
  // Partial Fisher-Yates shuffle: the first maxFeatures entries are a
  // uniform random subset without replacement.
  for (let i = 0; i < maxFeatures; i++) {
    const j = i + Math.floor(rng() * (d - i));
    [all[i], all[j]] = [all[j], all[i]];
  }
  return all.slice(0, maxFeatures);
}

function bestSplitOf(
  X: number[][],
  y: number[],
  rng: () => number,
  maxFeatures: number | null,
): { feature: number; threshold: number } | null {
  const n = y.length;
  const features = pickFeatures(rng, X[0].length, maxFeatures);
  let best: { feature: number; threshold: number } | null = null;
  let bestScore = giniImpurity(y) - 1e-12;
  for (const j of features) {
    const vals = Array.from(new Set(X.map((row) => row[j]))).sort((a, b) => a - b);
    for (let k = 0; k < vals.length - 1; k++) {
      const t = (vals[k] + vals[k + 1]) / 2;
      const yLeft: number[] = [];
      const yRight: number[] = [];
      for (let i = 0; i < n; i++) (X[i][j] <= t ? yLeft : yRight).push(y[i]);
      if (yLeft.length === 0 || yRight.length === 0) continue;
      const score =
        (yLeft.length * giniImpurity(yLeft) + yRight.length * giniImpurity(yRight)) / n;
      if (score < bestScore) {
        bestScore = score;
        best = { feature: j, threshold: t };
      }
    }
  }
  return best;
}

/** Grow an unlimited-depth CART tree; same node objects as Task 1. */
function trainTree(
  X: number[][],
  y: number[],
  rng: () => number,
  maxFeatures: number | null,
): TreeNode {
  if (giniImpurity(y) === 0) return { leaf: true, prediction: majorityClass(y) };
  const split = bestSplitOf(X, y, rng, maxFeatures);
  if (split === null) return { leaf: true, prediction: majorityClass(y) };
  const XL: number[][] = [];
  const yL: number[] = [];
  const XR: number[][] = [];
  const yR: number[] = [];
  for (let i = 0; i < y.length; i++) {
    if (X[i][split.feature] <= split.threshold) {
      XL.push(X[i]);
      yL.push(y[i]);
    } else {
      XR.push(X[i]);
      yR.push(y[i]);
    }
  }
  return {
    leaf: false,
    feature: split.feature,
    threshold: split.threshold,
    left: trainTree(XL, yL, rng, maxFeatures),
    right: trainTree(XR, yR, rng, maxFeatures),
  };
}

/** Predict every row of X with one tree. */
function treePredict(tree: TreeNode, X: number[][]): number[] {
  return X.map((x) => {
    let node = tree;
    while (!node.leaf)
      node = x[node.feature] <= node.threshold ? node.left : node.right;
    return node.prediction;
  });
}

// ---------------------------------------------------------------------------
// Core functions — YOU implement these three
// ---------------------------------------------------------------------------

/**
 * Draw a bootstrap sample: n indices from {0, …, n-1} WITH replacement.
 *
 * Returns an array of n integer indices (duplicates expected — that's the
 * point; ≈ 63.2% of the rows appear at least once).
 *
 * TODO: implement — n draws of Math.floor(rng() * n).
 */
function bootstrapSample(rng: () => number, n: number): number[] {
  // TODO: implement the bootstrap draw
  throw new Error("TODO: implement bootstrapSample()");
}

/**
 * Train a random forest: nTrees CART trees, EACH on its own bootstrap sample
 * of the rows, EACH restricted to maxFeatures random features per split (the
 * provided trainTree handles that part — just pass it through).
 *
 * Returns the array of nTrees TreeNode roots.
 *
 * TODO: implement.
 *   1. For each of the nTrees rounds, draw indices with your bootstrapSample
 *      and gather the corresponding rows of X and entries of y.
 *   2. Train a tree on that resample with the provided trainTree (forward
 *      rng and maxFeatures), and collect it.
 *   3. Return the array of trees.
 */
function trainForest(
  X: number[][],
  y: number[],
  rng: () => number,
  nTrees: number,
  maxFeatures: number | null,
): TreeNode[] {
  // TODO: implement the bootstrap-then-train loop
  throw new Error("TODO: implement trainForest()");
}

/**
 * Majority vote of the forest: predict X with every tree (treePredict),
 * then, per sample, output the class most trees chose.
 *
 * With binary {0,1} labels the vote reduces to: mean over trees ≥ 0.5 → 1.
 * (N_TREES is odd, so there are no exact ties.)
 *
 * Returns an array of X.length {0, 1} predictions.
 *
 * TODO: implement.
 *   1. Collect every tree's treePredict(tree, X) (an nTrees × N array).
 *   2. Take the per-sample vote per the rule above and return it.
 */
function forestPredict(trees: TreeNode[], X: number[][]): number[] {
  // TODO: implement the majority vote
  throw new Error("TODO: implement forestPredict()");
}

// ---------------------------------------------------------------------------
// Harness (provided — do not edit)
// ---------------------------------------------------------------------------

function acc(pred: number[], y: number[]): number {
  let correct = 0;
  for (let i = 0; i < y.length; i++) if (pred[i] === y[i]) correct++;
  return correct / y.length;
}

function main(): void {
  console.log("Task 2 — Bagging → random forest\n");

  const { XTrain, yTrain, XTest, yTest } = makeData();
  console.log(`  Data: ${yTrain.length} train / ${yTest.length} test, 5 features`);
  console.log(`  (2 signal + 3 pure noise), ${NOISE_FLIP * 100}% label noise\n`);

  // ── Bootstrap statistics ───────────────────────────────────────────────────
  console.log("[1/3] Bootstrap sampling (63.2% unique)...");
  let rng = makeRng(SEED);
  const n = yTrain.length;
  const uniqueFracs: number[] = [];
  let allHaveDups = true;
  for (let s = 0; s < 20; s++) {
    const idx = bootstrapSample(rng, n);
    const uniq = new Set(idx).size;
    uniqueFracs.push(uniq / n);
    if (uniq === n) allHaveDups = false;
  }
  const meanUnique = uniqueFracs.reduce((a, b) => a + b, 0) / uniqueFracs.length;
  console.log(
    `  20 bootstraps of n=${n}: mean unique fraction = ${meanUnique.toFixed(4)}`,
  );
  console.log(
    `  (theory: 1 - 1/e ≈ 0.6321) — every sample has duplicates: ${allHaveDups}\n`,
  );

  // ── Baseline: one deep tree on everything ──────────────────────────────────
  console.log("[2/3] Baseline: single deep tree (all rows, all features)...");
  rng = makeRng(SEED);
  const baseline = trainTree(XTrain, yTrain, rng, null);
  const baseTrain = acc(treePredict(baseline, XTrain), yTrain);
  const baseTest = acc(treePredict(baseline, XTest), yTest);
  console.log(
    `  train acc = ${baseTrain.toFixed(4)}   test acc = ${baseTest.toFixed(4)}\n`,
  );

  // ── The forest ─────────────────────────────────────────────────────────────
  console.log(`[3/3] Random forest (${N_TREES} trees, maxFeatures=${MAX_FEATURES})...`);
  rng = makeRng(SEED);
  const forest = trainForest(XTrain, yTrain, rng, N_TREES, MAX_FEATURES);
  const treeAccs = forest.map((t) => acc(treePredict(t, XTest), yTest));
  const ensTest = acc(forestPredict(forest, XTest), yTest);
  const meanTreeAcc = treeAccs.reduce((a, b) => a + b, 0) / treeAccs.length;
  console.log(
    `  individual tree test accs (first 8): [${treeAccs
      .slice(0, 8)
      .map((a) => a.toFixed(3))
      .join(", ")}]`,
  );
  console.log(
    `  individual: mean = ${meanTreeAcc.toFixed(4)}  ` +
      `min = ${Math.min(...treeAccs).toFixed(4)}  max = ${Math.max(...treeAccs).toFixed(4)}`,
  );
  console.log(`  ensemble (majority vote) test acc = ${ensTest.toFixed(4)}`);

  // ── Acceptance checks ──────────────────────────────────────────────────────
  console.log("\nAcceptance:");
  const okBoot = allHaveDups && meanUnique >= 0.55 && meanUnique <= 0.72;
  const okVsMean = ensTest >= meanTreeAcc;
  const okVsBase = ensTest >= baseTest;
  console.log(
    `  [${okBoot ? "x" : " "}] bootstraps have duplicates; unique fraction ≈ 0.632 (got ${meanUnique.toFixed(4)})`,
  );
  console.log(
    `  [${okVsMean ? "x" : " "}] ensemble ≥ mean individual tree  (${ensTest.toFixed(4)} ≥ ${meanTreeAcc.toFixed(4)})`,
  );
  console.log(
    `  [${okVsBase ? "x" : " "}] ensemble ≥ single deep-tree baseline  (${ensTest.toFixed(4)} ≥ ${baseTest.toFixed(4)})`,
  );

  if (okBoot && okVsMean && okVsBase) {
    console.log("\n  All acceptance checks passed.");
  } else {
    console.log("\n  Some checks failed — revisit your implementation.");
  }
}

main();

// This file has no imports; the empty export marks it as an ES module so its
// top-level names stay file-scoped (keeps `pnpm typecheck` happy across modules).
export {};

File: modules/01e-trees-ensembles/py/02_random_forest.py
"""
Task 2 🟡 — Bagging → random forest.

What you'll learn:
  - Bootstrap sampling: draw n indices WITH replacement — each resample leaves
    out ≈ 36.8% of the rows (the fraction of unique rows → 1 - 1/e ≈ 63.2%).
  - Bagging: train one tree per bootstrap, average (majority-vote) their
    predictions — averaging B noisy estimators divides the uncorrelated part
    of their variance by B.
  - The random-forest trick: at EVERY split, consider only a random subset of
    max_features features, so the trees stop all making the same greedy first
    split and become decorrelated — which is what makes the averaging work.

The math (README derives each step):

  Unique-fraction of a bootstrap:  P(row i never drawn in n tries)
      = (1 - 1/n)^n → e⁻¹ ≈ 0.368,  so  unique ≈ 63.2% of rows.

  Variance of an average of B estimators with variance σ² and pairwise
  correlation ρ:
      Var( (1/B) Σ f_b ) = ρσ² + (1-ρ)σ²/B
  Bagging shrinks the second term; feature subsampling shrinks ρ — the first.

You implement: bootstrap_sample, train_forest, forest_predict. A full
single-tree trainer that accepts max_features (the per-split feature
sampler), the dataset (2 signal + 3 pure-noise features), and the harness
comparing individual trees vs the ensemble vs a single deep-tree baseline
are provided and runnable.

How to run:
  uv run python modules/01e-trees-ensembles/py/02_random_forest.py
"""

from __future__ import annotations

import numpy as np

SEED = 21
N = 520  # total samples (first N_TRAIN train, rest test)
N_TRAIN = 370
NOISE_FLIP = 0.10
N_TREES = 25
MAX_FEATURES = 2  # features considered per split (out of 5)


# ---------------------------------------------------------------------------
# Synthetic data  (provided — do not edit)
# ---------------------------------------------------------------------------


def make_data() -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """
    5 features: x0, x1 carry the signal (the offset-XOR pattern from Task 1);
    x2..x4 are pure Gaussian noise — bait for an overfitting deep tree.
    NOISE_FLIP of the labels are flipped.

    Returns: X_train (N_TRAIN, 5), y_train, X_test, y_test.
    """
    rng = np.random.default_rng(SEED)
    X_signal = rng.uniform(-3.0, 3.0, size=(N, 2))
    X_noise = rng.normal(0.0, 1.0, size=(N, 3))
    X = np.hstack([X_signal, X_noise])
    y = ((X[:, 0] > 0.8) ^ (X[:, 1] > -0.7)).astype(int)
    flip = rng.random(N) < NOISE_FLIP
    y = np.where(flip, 1 - y, y)
    return X[:N_TRAIN], y[:N_TRAIN], X[N_TRAIN:], y[N_TRAIN:]


# ---------------------------------------------------------------------------
# Single-tree trainer  (provided — do not edit)
#
# The same CART you built in Task 1, with ONE addition: if max_features is
# set, each split only scans a random subset of that many features (drawn
# with the rng) — the random-forest decorrelation trick.
# ---------------------------------------------------------------------------


def _gini(y: np.ndarray) -> float:
    if y.size == 0:
        return 0.0
    _, counts = np.unique(y, return_counts=True)
    p = counts / y.size
    return float(1.0 - np.sum(p**2))


def _best_split(
    X: np.ndarray,
    y: np.ndarray,
    rng: np.random.Generator,
    max_features: int | None,
) -> tuple[int, float] | None:
    n, d = X.shape
    if max_features is None:
        features = np.arange(d)
    else:
        features = rng.choice(d, size=min(max_features, d), replace=False)
    best: tuple[int, float] | None = None
    best_score = _gini(y) - 1e-12
    for j in features:
        vals = np.unique(X[:, j])
        for k in range(len(vals) - 1):
            t = (vals[k] + vals[k + 1]) / 2.0
            left = X[:, j] <= t
            n_l = int(left.sum())
            if n_l == 0 or n_l == n:
                continue
            score = (n_l * _gini(y[left]) + (n - n_l) * _gini(y[~left])) / n
            if score < best_score:
                best_score = score
                best = (int(j), float(t))
    return best


def train_tree(
    X: np.ndarray,
    y: np.ndarray,
    rng: np.random.Generator,
    max_features: int | None = None,
) -> dict:
    """Grow an unlimited-depth CART tree; same node dicts as Task 1."""
    if _gini(y) == 0.0:
        return {"leaf": True, "prediction": int(np.bincount(y).argmax())}
    split = _best_split(X, y, rng, max_features)
    if split is None:
        return {"leaf": True, "prediction": int(np.bincount(y).argmax())}
    j, t = split
    left = X[:, j] <= t
    return {
        "leaf": False,
        "feature": j,
        "threshold": t,
        "left": train_tree(X[left], y[left], rng, max_features),
        "right": train_tree(X[~left], y[~left], rng, max_features),
    }


def tree_predict(tree: dict, X: np.ndarray) -> np.ndarray:
    """Predict every row of X with one tree."""
    out = np.empty(len(X), dtype=int)
    for i, x in enumerate(X):
        node = tree
        while not node["leaf"]:
            node = node["left"] if x[node["feature"]] <= node["threshold"] else node["right"]
        out[i] = node["prediction"]
    return out


# ---------------------------------------------------------------------------
# Core functions — YOU implement these three
# ---------------------------------------------------------------------------


def bootstrap_sample(rng: np.random.Generator, n: int) -> np.ndarray:
    """
    Draw a bootstrap sample: n indices from {0, …, n-1} WITH replacement.

    Returns: (n,) int array of row indices (duplicates expected — that's the
    point; ≈ 63.2% of the rows appear at least once).

    TODO: implement — one call to rng.integers with the right bounds and size.
    """
    # TODO: implement the bootstrap draw
    raise NotImplementedError("TODO: implement bootstrap_sample()")


def train_forest(
    X: np.ndarray,
    y: np.ndarray,
    rng: np.random.Generator,
    n_trees: int,
    max_features: int | None,
) -> list[dict]:
    """
    Train a random forest: n_trees CART trees, EACH on its own bootstrap
    sample of the rows, EACH restricted to max_features random features per
    split (the provided train_tree handles that part — just pass it through).

    Returns: list of n_trees tree dicts.

    TODO: implement.
      1. For each of the n_trees rounds, draw indices with your
         bootstrap_sample and slice X and y by them.
      2. Train a tree on that resample with the provided train_tree
         (forward rng and max_features), and collect it.
      3. Return the list of trees.
    """
    # TODO: implement the bootstrap-then-train loop
    raise NotImplementedError("TODO: implement train_forest()")


def forest_predict(trees: list[dict], X: np.ndarray) -> np.ndarray:
    """
    Majority vote of the forest: predict X with every tree (tree_predict),
    then, per sample, output the class most trees chose.

    With binary {0,1} labels the vote reduces to: mean over trees ≥ 0.5 → 1.
    (N_TREES is odd, so there are no exact ties.)

    Returns: (len(X),) int array of {0, 1} predictions.

    TODO: implement.
      1. Stack every tree's tree_predict(tree, X) into an (n_trees, N) matrix.
      2. Take the per-column vote per the rule above and return it as ints.
    """
    # TODO: implement the majority vote
    raise NotImplementedError("TODO: implement forest_predict()")


# ---------------------------------------------------------------------------
# Harness  (provided — do not edit)
# ---------------------------------------------------------------------------


def main() -> None:
    print("Task 2 — Bagging → random forest\n")

    X_train, y_train, X_test, y_test = make_data()
    print(f"  Data: {len(y_train)} train / {len(y_test)} test, 5 features")
    print(f"  (2 signal + 3 pure noise), {NOISE_FLIP:.0%} label noise\n")

    # ── Bootstrap statistics ─────────────────────────────────────────────────
    print("[1/3] Bootstrap sampling (63.2% unique)...")
    rng = np.random.default_rng(SEED)
    n = len(y_train)
    unique_fracs = []
    all_have_dups = True
    for _ in range(20):
        idx = bootstrap_sample(rng, n)
        uniq = len(np.unique(idx))
        unique_fracs.append(uniq / n)
        if uniq == n:
            all_have_dups = False
    mean_unique = float(np.mean(unique_fracs))
    print(f"  20 bootstraps of n={n}: mean unique fraction = {mean_unique:.4f}")
    print(f"  (theory: 1 - 1/e ≈ 0.6321) — every sample has duplicates: {all_have_dups}\n")

    # ── Baseline: one deep tree on everything ────────────────────────────────
    print("[2/3] Baseline: single deep tree (all rows, all features)...")
    rng = np.random.default_rng(SEED)
    baseline = train_tree(X_train, y_train, rng, max_features=None)
    base_train = float(np.mean(tree_predict(baseline, X_train) == y_train))
    base_test = float(np.mean(tree_predict(baseline, X_test) == y_test))
    print(f"  train acc = {base_train:.4f}   test acc = {base_test:.4f}\n")

    # ── The forest ───────────────────────────────────────────────────────────
    print(f"[3/3] Random forest ({N_TREES} trees, max_features={MAX_FEATURES})...")
    rng = np.random.default_rng(SEED)
    forest = train_forest(X_train, y_train, rng, N_TREES, MAX_FEATURES)
    tree_accs = np.array([float(np.mean(tree_predict(t, X_test) == y_test)) for t in forest])
    ens_test = float(np.mean(forest_predict(forest, X_test) == y_test))
    print(f"  individual tree test accs (first 8): {np.round(tree_accs[:8], 3)}")
    print(
        f"  individual: mean = {tree_accs.mean():.4f}  "
        f"min = {tree_accs.min():.4f}  max = {tree_accs.max():.4f}"
    )
    print(f"  ensemble (majority vote) test acc = {ens_test:.4f}")

    # ── Acceptance checks ────────────────────────────────────────────────────
    print("\nAcceptance:")
    ok_boot = all_have_dups and 0.55 <= mean_unique <= 0.72
    ok_vs_mean = ens_test >= float(tree_accs.mean())
    ok_vs_base = ens_test >= base_test
    print(
        f"  [{'x' if ok_boot else ' '}] bootstraps have duplicates; unique fraction "
        f"≈ 0.632 (got {mean_unique:.4f})"
    )
    print(
        f"  [{'x' if ok_vs_mean else ' '}] ensemble ≥ mean individual tree  "
        f"({ens_test:.4f} ≥ {tree_accs.mean():.4f})"
    )
    print(
        f"  [{'x' if ok_vs_base else ' '}] ensemble ≥ single deep-tree baseline  "
        f"({ens_test:.4f} ≥ {base_test:.4f})"
    )

    if ok_boot and ok_vs_mean and ok_vs_base:
        print("\n  All acceptance checks passed.")
    else:
        print("\n  Some checks failed — revisit your implementation.")


if __name__ == "__main__":
    main()
